### PR TITLE
Chore/refactor tests

### PR DESCRIPTION
**Pull request by Aboudjem**
- Refactor test architecture and rename files.
- Remove unused functions
- Fix Solidity compiler warning.
- Improve comments, and formatting.
- Fix invariants tests
- Fix Fuzz Tests
- Reformat Events / Errors file on tests
- Rename `preparePackedUserOperation` to `buildPackedUserOperation`.
- Move functions to helpers.
- Refactor helpers and rename functions.
- Remove unused code
- Update test contracts for consistency.
- Add negative tests.
- Add prefunding for smart account tests
- Add inline comments.
- Add natspec comments.
- Improve comments and update Solidity version
- Add check on `onUninstall` hook which was causing issues on invariant

### CODE REVIEW

**Comment by livingrockrises** [comment]
why is this increasing?


Nexus | upgradeToAndCall | +2,929 ❌ | +230.99%
-- | -- | -- | --

**Comment by Aboudjem** [comment]
> why is this increasing?
> 
> Nexus	upgradeToAndCall	+2,929 ❌	+230.99%

because of more tests + different scenarios (some more complex)

File: contracts/mocks/MockValidator.sol [hunk 1]
```diff
@@ -28,11 +29,13 @@ contract MockValidator is IValidator {
     }
 
     function onInstall(bytes calldata data) external {
+        require(IModuleManager(msg.sender).isModuleInstalled(MODULE_TYPE_VALIDATOR, address(this), ""), "Validator is still installed");
```
Comment (livingrockrises) [inline contracts/mocks/MockValidator.sol:32]: I should be able to call onInstall again on already installed module right?
for example K1 validator I might want to swap an owner

Comment (Aboudjem) [reply contracts/mocks/MockValidator.sol:32]: The `onInstall` function should only be called within an `installModule`. It's a mock, but we need to consider the possibility that `onInstall` could be called directly. We should think about how to secure it or call it properly without causing any issues so I think it's fine here as MockValidator is not meant to be used on prod

Comment (livingrockrises) [reply contracts/mocks/MockValidator.sol:32]: even if there's a real K1 validator (check one of the modules from module-kit which are being audited) no one would write two different methods onInstall and transferOwnership I feel. if it's directly called it's always referred by msg.sender so it shouldn't be an issue. imo this should be removed. test should check onInstall was called for smart account or some random sender

File: contracts/mocks/MockValidator.sol [hunk 2]
```diff
@@ -28,11 +29,13 @@ contract MockValidator is IValidator {
...
         smartAccountOwners[msg.sender] = address(bytes20(data));
     }
 
     function onUninstall(bytes calldata data) external {
         data;
+        require(!IModuleManager(msg.sender).isModuleInstalled(MODULE_TYPE_VALIDATOR, address(this), ""), "Validator is still installed");
```
Comment (livingrockrises) [inline contracts/mocks/MockValidator.sol:38]: same I feel for onUninstall 
how much gas does this action increase?
this will also affect smart account deployment cost.

File: scripts/foundry/Deploy.s.sol [hunk 3]
```diff
@@ -1,5 +1,5 @@
 // SPDX-License-Identifier: UNLICENSED
-pragma solidity >=0.8.0 <0.9.0;
+pragma solidity >=0.8.24 <0.9.0;
```
Comment (livingrockrises) [inline scripts/foundry/Deploy.s.sol:2]: why this change only at one place

File: test/foundry/fork/arbitrum/ArbitrumSmartAccountUpgradeTest.t.sol [hunk 4]
```diff
@@ -61,95 +126,32 @@ contract Test_SmartAccountV2toV3 is SmartAccountTestLab, ArbitrumForkSettings {
...
-        PackedUserOperation[] memory userOps = preparePackedUserOperation(BOB, Nexus(payable(address(smartAccountV2))), EXECTYPE_DEFAULT, execution);
-
-        // Execute the operation via the EntryPoint
-        ENTRYPOINT_V_0_7.handleOps(userOps, payable(OWNER_ADDRESS));
-
-        // Check the recipient received the ETH
-        assertEq(address(recipient).balance, amount, "ETH transfer failed");
-    }
-
-    function preparePackedUserOperation(
```
Comment (livingrockrises) [inline test/foundry/fork/arbitrum/ArbitrumSmartAccountUpgradeTest.t.sol:147]: what was wrong with this name?

File: test/foundry/integration/UpgradeSmartAccountTest.t.sol [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        Nexus newSmartAccount = new Nexus();
+        bytes memory callData = abi.encodeWithSelector(Nexus.upgradeToAndCall.selector, address(newSmartAccount), "");
+
+        Execution[] memory execution = new Execution[](1);
+        execution[0] = Execution(address(BOB_ACCOUNT), 0, callData);
+
+        PackedUserOperation[] memory userOps = buildPackedUserOperation(BOB, BOB_ACCOUNT, EXECTYPE_DEFAULT, execution, address(VALIDATOR_MODULE));
+        ENTRYPOINT.handleOps(userOps, payable(address(BOB.addr)));
+
+        address newImplementation = BOB_ACCOUNT.getImplementation();
```
Comment (livingrockrises) [inline test/foundry/integration/UpgradeSmartAccountTest.t.sol:36]: you mean this is failing currently?

File: test/foundry/invariant/ActorManagerInvariantTest.t.sol [hunk 6]
```diff
@@ -1,43 +1,42 @@
...
-contract ActorManager is BaseInvariantTest {
+import "../utils/Imports.sol";
+import { InvariantBaseTest } from "./base/InvariantBaseTest.t.sol";
+import { ExecutionHandlerTest } from "./handlers/ExecutionHandlerTest.t.sol";
+import { AccountCreationHandlerTest } from "./handlers/AccountCreationHandlerTest.t.sol";
+import { ModuleManagementHandlerTest } from "./handlers/ModuleManagementHandlerTest.t.sol";
+import { DepositManagementHandlerTest } from "./handlers/DepositManagementHandlerTest.t.sol";
+
+/// @title ActorManagerInvariantTest
+/// @notice Coordinates test actions across different actors using handlers.
```
Comment (livingrockrises) [inline test/foundry/invariant/ActorManagerInvariantTest.t.sol:12]: can you describe more about actors here and everywhere

File: test/foundry/invariant/handlers/DepositManagementHandlerTest.t.sol [hunk 7]
```diff
@@ -1,55 +1,72 @@
...
-    // Handles a deposit operation while verifying that the state remains consistent
+    /// @notice Handles a deposit operation while verifying that the state remains consistent.
+    /// @param amount The amount to deposit.
     function invariant_handleDeposit(uint256 amount) public {
         Execution[] memory executions = new Execution[](1);
         executions[0] = Execution({ target: address(nexusAccount), value: amount, callData: abi.encodeWithSignature("addDeposit()") });
 
         // Execute deposit through the ENTRYPOINT with invariant checks
-        PackedUserOperation[] memory userOps = preparePackedUserOperation(signer, nexusAccount, EXECTYPE_DEFAULT, executions);
+        PackedUserOperation[] memory userOps = buildPackedUserOperation(
```
Comment (livingrockrises) [inline test/foundry/invariant/handlers/DepositManagementHandlerTest.t.sol:28]: some places it's replaced with buildPackedUserOperation, whereas some place it's replaced with buildUserOperation

Comment (Aboudjem) [reply test/foundry/invariant/handlers/DepositManagementHandlerTest.t.sol:28]: `buildUserOperation` is only on Arbitrum fork testing, as it is used to create a v0.6 compatible userOp

File: contracts/mocks/MockValidator.sol (hunk 2 above)
Comment (livingrockrises) [reply contracts/mocks/MockValidator.sol:38]: for stateful validator it would always be referred by msg.sender address

File: test/foundry/integration/UpgradeSmartAccountTest.t.sol (hunk 5 above)
Comment (Aboudjem) [reply test/foundry/integration/UpgradeSmartAccountTest.t.sol:36]: No this one from v3 works perfectly, but on fork while upgrading a v2 like here it's failing that's why I have commented the test:

Screenshot 2024-05-20 at 7 11 43 PM

File: scripts/foundry/Deploy.s.sol (hunk 3 above)
Comment (Aboudjem) [reply scripts/foundry/Deploy.s.sol:2]: I just replaced all 0.8.0 by 0.8.24

Comment (livingrockrises) [reply scripts/foundry/Deploy.s.sol:2]: I thought it was ^0.8.24 already for all files

you mean all files have
pragma solidity >=0.8.24 <0.9.0; now

File: test/foundry/fork/arbitrum/ArbitrumSmartAccountUpgradeTest.t.sol (hunk 4 above)
Comment (Aboudjem) [reply test/foundry/fork/arbitrum/ArbitrumSmartAccountUpgradeTest.t.sol:147]: nothing, we had many different namings `prepare`, `build`, `create`, had to choose one for consistency

File: scripts/foundry/Deploy.s.sol (hunk 3 above)
Comment (Aboudjem) [reply scripts/foundry/Deploy.s.sol:2]: this was just replace accross all files, but this one had different pragma I'll revert it

File: contracts/mocks/MockValidator.sol (hunk 1 above)
Comment (Aboudjem) [reply contracts/mocks/MockValidator.sol:32]: So, we acknowledge this and accept the potential risk of breaking a Smart Account if it's mistakenly used.

Comment (livingrockrises) [reply contracts/mocks/MockValidator.sol:32]: as discussed on mock validator it should be fine. but we should address it (maybe chat with Rhinestone also?) before we start K1 validator in prod on deployed accounts.

File: scripts/foundry/Deploy.s.sol (hunk 3 above)
Comment (livingrockrises) [reply scripts/foundry/Deploy.s.sol]: got it

File: test/foundry/fork/arbitrum/ArbitrumSmartAccountUpgradeTest.t.sol (hunk 4 above)
Comment (livingrockrises) [reply test/foundry/fork/arbitrum/ArbitrumSmartAccountUpgradeTest.t.sol:147]: I liked prepare but should be fine. 
good if we replace such things in final version to avoid conflict
can focus on fixing tests, descriptions or solhint rules and slither in these prs.

File: test/foundry/integration/UpgradeSmartAccountTest.t.sol (hunk 5 above)
Comment (livingrockrises) [reply test/foundry/integration/UpgradeSmartAccountTest.t.sol:36]: yes yes I meant after the upgrade. 
perhaps we have to remove getImplementation from the account 
larger impact to investigate would be by testing 
v2 -> V3 -> V4(some other implementation) 
cause 
storage slot for V2 is address of self
storage slot for V3 is 1967_SLOT

File: test/foundry/invariant/handlers/DepositManagementHandlerTest.t.sol (hunk 7 above)
Comment (livingrockrises) [reply test/foundry/invariant/handlers/DepositManagementHandlerTest.t.sol:28]: then we should go with build/prepare useropeartion/packeduseroperation

**Comment by livingrockrises** [comment]
why does this also increase?

AccountFactory | withdrawTo | +13,803 ❌ | +95.46%
-- | -- | -- | --

File: contracts/mocks/MockValidator.sol (hunk 1 above)
Comment (livingrockrises) [reply contracts/mocks/MockValidator.sol:32]: maybe this too for tomorrow's call
